src/commands/logs/list.rs

```diff
@@ -143,10 +168,327 @@ fn execute_single_fetch(
     Ok(())
 }
 
+/// Manages deduplication of log entries with a bounded buffer
+struct LogDeduplicator {
+    /// Set of seen log IDs for quick lookup
+    seen_ids: HashSet<String>,
+    /// Buffer of log entries in order (for maintaining size limit)
+    buffer: VecDeque<LogEntry>,
+    /// Maximum size of the buffer
+    max_size: usize,
+}
+
+const MAX_DEDUP_BUFFER_SIZE: usize = 10_000;
+
+impl LogDeduplicator {
+    fn new(max_size: usize) -> Self {
+        Self {
+            seen_ids: HashSet::new(),
+            buffer: VecDeque::new(),
+            max_size,
+        }
+    }
+
+    /// Add new logs and return only the ones that haven't been seen before
+    fn add_logs(&mut self, new_logs: Vec<LogEntry>) -> Vec<LogEntry> {
+        let mut unique_logs = Vec::new();
+
+        for log in new_logs {
+            if !self.seen_ids.contains(&log.item_id) {
+                self.seen_ids.insert(log.item_id.clone());
+                self.buffer.push_back(log.clone());
+                unique_logs.push(log);
+            }
+        }
+
+        // Maintain buffer size limit by removing oldest entries
+        while self.buffer.len() > self.max_size {
+            if let Some(removed_log) = self.buffer.pop_front() {
+                self.seen_ids.remove(&removed_log.item_id);
+            }
+        }
+
+        unique_logs
+    }
+}
```

Comment:
[change request] If I am understanding the intention behind the code correctly, I think what you want here is an LRU cache, which is implemented in the [`lru` crate](https://crates.io/crates/lru).

The current implementation with the separate `HashSet` and `VecDeque` is brittle. Consider, for instance, the case where the same log is infinitely repeated. The first 10k instances are dedeuplicated, but then, the `VecDeque` is full, so on every iteration, we pop the log from the `VecDeque` and remove it from the `HashSet` (even though it is the only item in the `HashSet`). As a result, from the 10001-th iteration onwards, we print the log every time.